compiler/rustc_middle/src/dep_graph/dep_node.rs

```diff
@@ -350,14 +352,16 @@ impl DepNodeExt for DepNode {
     fn from_label_string(label: &str, def_path_hash: DefPathHash) -> Result<DepNode, ()> {
         let kind = dep_kind_from_label_string(label)?;
 
-        if !kind.can_reconstruct_query_key() {
-            return Err(());
-        }
-
-        if kind.has_params {
-            Ok(DepNode::from_def_path_hash(def_path_hash, kind))
-        } else {
-            Ok(DepNode::new_no_params(kind))
+        match kind.fingerprint_style() {
+            FingerprintStyle::Opaque => Err(()),
+            FingerprintStyle::Unit => {
+                if !kind.has_params {
```

Comment:
`FingerprintStyle::Unit` should now be equivalent to `has_params == false`, isn't it? We should be able to get rid of `has_params`.

Reply:
Not currently. A query with `()` key currently "has_params"; the only two things to not have params are Null and TraitSelect which are not generated through the normal macro infrastructure. dropped the "fix" for that from this branch for now, though I can re-add it, it's not big (b2b53bb9a4f2f5e1f96f8f01ad7b378f5eb1b4d1).